File: apps/android/app/src/main/java/com/kungsbackacarcommunity/app/notifications/FirebaseNotificationsRepository.kt
package com.kungsbackacarcommunity.app.notifications

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.Query
import com.google.firebase.functions.FirebaseFunctions
import com.kungsbackacarcommunity.app.firebase.awaitOrThrow
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.combine

/**
 * [NotificationsRepository] backed by an owner-only Firestore listener on the
 * inbox plus the notifications.markRead / markAllRead / delete / deleteAll
 * callables (europe-west1), Phase 12 slice 21. The listener is bounded to the newest
 * [Notifications.INBOX_QUERY_LIMIT] items (createdAt descending — a
 * single-field orderBy, so Firestore's automatic index suffices); items are
 * additionally sorted newest-first client-side ([Notifications.sortedForInbox])
 * with the same ordering. Guarded ([createIfAvailable]).
 */
class FirebaseNotificationsRepository private constructor(
    private val firestore: FirebaseFirestore,
    private val functions: FirebaseFunctions,
) : NotificationsRepository {

    override fun observeNotifications(uid: String): Flow<NotificationsState> = callbackFlow {
        val registration =
            firestore
                .collection(NOTIFICATIONS)
                .document(uid)
                .collection(ITEMS)
                .orderBy(CREATED_AT, Query.Direction.DESCENDING)
                .limit(Notifications.INBOX_QUERY_LIMIT)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        trySend(NotificationsState.Error)
                        return@addSnapshotListener
                    }
                    val items = snapshot?.documents?.mapNotNull { it.toNotification() } ?: emptyList()
                    trySend(NotificationsState.Loaded(Notifications.sortedForInbox(items)))
                }
        awaitClose { registration.remove() }
    }

    override fun observeUnread(uid: String): Flow<Boolean> {
        // Two cheap listeners, both bound while the dot is on screen: the newest
        // inbox item (a limit(1) createdAt-DESC query, the same automatic index
        // observeNotifications uses) and the caller's userPrivate last-seen
        // marker. The dot lights when the newest notification post-dates the
        // marker — the exact mirror of the community chat dot, and lighter than
        // the old full-inbox listener the aggregate used to run.
        val newestCreatedAt: Flow<Long?> = callbackFlow {
            val registration =
                firestore
                    .collection(NOTIFICATIONS)
                    .document(uid)
                    .collection(ITEMS)
                    .orderBy(CREATED_AT, Query.Direction.DESCENDING)
                    .limit(1)
                    .addSnapshotListener { snapshot, error ->
                        if (error != null) {
                            if ((error as? FirebaseFirestoreException)?.code ==
                                FirebaseFirestoreException.Code.PERMISSION_DENIED
                            ) {
                                // Access revoked (account restriction / rules
                                // gating): hard-clear the dot even if a stale
                                // cached snapshot is present — never keep it lit
                                // for an inbox the user can no longer read.
                                // Mirrors CommunityChatRepository.observeUnread.
                                trySend(null)
                                return@addSnapshotListener
                            }
                            // Other (transient) error with no cached data: keep
                            // the last-known value rather than emitting a
                            // misleading no-unread. With cached data, fall
                            // through and use it.
                            if (snapshot == null) return@addSnapshotListener
                        }
                        trySend(
                            snapshot?.documents?.firstOrNull()
                                ?.getTimestamp(CREATED_AT)?.toDate()?.time,
                        )
                    }
            awaitClose { registration.remove() }
        }
        val lastSeenAt: Flow<Long?> = callbackFlow {
            val registration =
                firestore
                    .collection(USER_PRIVATE)
                    .document(uid)
                    .addSnapshotListener { snapshot, error ->
                        if (error != null && snapshot == null) {
                            // Transient failure with no cached marker: keep the
                            // last-known marker rather than momentarily reading it
                            // as missing, which could wrongly re-light the dot.
                            return@addSnapshotListener
                        }
                        trySend(snapshot?.getTimestamp(LAST_SEEN_AT)?.toDate()?.time)
                    }
            awaitClose { registration.remove() }
        }
        return combine(newestCreatedAt, lastSeenAt) { newest, marker ->
            Notifications.hasUnread(newest, marker)
        }
    }

    override suspend fun markRead(notificationId: String) {
        call(MARK_READ, mapOf("notificationId" to notificationId))
    }

    override suspend fun markAllRead() {
        call(MARK_ALL_READ, emptyMap())
    }

    override suspend fun markSeen() {
        call(MARK_SEEN, emptyMap())
    }

    override suspend fun deleteNotification(notificationId: String) {
        call(DELETE, mapOf("notificationId" to notificationId))
    }

    override suspend fun deleteAll() {
        call(DELETE_ALL, emptyMap())
    }

    private suspend fun call(name: String, data: Map<String, Any>) {
        functions.getHttpsCallable(name).call(data)
            .awaitOrThrow { "$name failed without a cause" }
    }

    companion object {
        private const val NOTIFICATIONS = "notifications"
        private const val ITEMS = "items"
        private const val CREATED_AT = "createdAt"
        private const val USER_PRIVATE = "userPrivate"
        private const val LAST_SEEN_AT = "notificationsLastSeenAt"
        private const val REGION = "europe-west1"
        private const val MARK_READ = "notifications-markRead"
        private const val MARK_ALL_READ = "notifications-markAllRead"
        private const val MARK_SEEN = "notifications-markSeen"
        private const val DELETE = "notifications-delete"
        private const val DELETE_ALL = "notifications-deleteAll"

        fun createIfAvailable(context: Context): NotificationsRepository? {
            if (FirebaseApp.getApps(context).isEmpty()) return null
            return FirebaseNotificationsRepository(
                FirebaseFirestore.getInstance(),
                FirebaseFunctions.getInstance(REGION),
            )
        }
    }
}

private fun DocumentSnapshot.toNotification(): AppNotification? {
    if (!exists()) return null
    val title = getString("title") ?: return null
    return AppNotification(
        id = id,
        category = NotificationCategory.fromWire(getString("category")),
        title = title,
        previewText = getString("previewText"),
        body = getString("body"),
        isRead = getBoolean("read") ?: false,
        createdAtMillis = getTimestamp("createdAt")?.toDate()?.time,
        // Both are written by every producer (notifications-core defaults them
        // to 'none'/null), but older inbox rows predate nothing here — the
        // fields have always been persisted. An unknown/absent actionType
        // degrades to NONE, i.e. a plain non-actionable row.
        actionType = NotificationActionType.fromWire(getString("actionType")),
        relatedEntityId = getString("relatedEntityId"),
    )
}

File: apps/ios/KCC/Notifications/FirebaseNotificationsRepository.swift
import FirebaseAuth
import FirebaseCore
import FirebaseFirestore
import Foundation

/// ``NotificationsRepository`` backed by Cloud Firestore + the notifications
/// callables — the iOS port of Android's `FirebaseNotificationsRepository`.
///
/// The inbox is an owner-only listener on `notifications/{uid}/items`, bounded
/// to the newest ``Notifications/inboxQueryLimit`` items (createdAt
/// descending — a single-field orderBy, so Firestore's automatic index
/// suffices); items are additionally sorted newest-first client-side
/// (``Notifications/sortedForInbox(_:)``). Read-state changes go through the
/// `notifications-markRead` / `-markAllRead` / `-markSeen` callables
/// (europe-west1, via ``KccFunctionsClient``), because the inbox is
/// backend-write-only.
///
/// Listener failures surface as ``NotificationsSnapshot/failed(code:)``
/// carrying the bare Firestore status name (reusing
/// ``FirebaseEventsRepository/firestoreStatusName(_:)``), never a silently
/// empty list — the same rule the events/garage listeners follow.
///
/// Construction is guarded (``createIfAvailable()`` returns nil without
/// Firebase config), mirroring ``FirebaseEventsRepository``.
///
/// Push delivery + token registration are OUT OF SCOPE: this type never
/// registers an APNs token or talks to FCM.
final class FirebaseNotificationsRepository: NotificationsRepository, @unchecked Sendable {
    private let firestore: Firestore
    private let functions: KccFunctionsClient

    private init(firestore: Firestore, functions: KccFunctionsClient) {
        self.firestore = firestore
        self.functions = functions
    }

    func notifications(uid: String) -> AsyncStream<NotificationsSnapshot> {
        let query =
            firestore
            .collection(Self.notificationsCollection)
            .document(uid)
            .collection(Self.itemsCollection)
            .order(by: Self.createdAtField, descending: true)
            .limit(to: Notifications.inboxQueryLimit)
        return AsyncStream { continuation in
            let registration = query.addSnapshotListener { snapshot, error in
                if let error {
                    // Bare status name only (never the exception text, which
                    // embeds the failing query and the project id) — see
                    // NotificationsSnapshot.failed.
                    continuation.yield(.failed(code: FirebaseEventsRepository.firestoreStatusName(error)))
                    return
                }
                let items = (snapshot?.documents ?? []).compactMap(Self.notification(from:))
                continuation.yield(.loaded(Notifications.sortedForInbox(items)))
            }
            let box = ListenerBox(registration: registration)
            continuation.onTermination = { _ in
                box.registration.remove()
            }
        }
    }

    func unread(uid: String) -> AsyncStream<Bool> {
        // Two cheap listeners, both bound while the dot is on screen: the
        // newest inbox item (a limit(1) createdAt-DESC query, the same
        // automatic index the inbox uses) and the caller's userPrivate
        // last-seen marker. The dot lights when the newest notification
        // post-dates the marker — Android's `observeUnread`.
        AsyncStream { continuation in
            let box = UnreadListenerBox()

            func emit() {
                continuation.yield(
                    Notifications.hasUnread(newest: box.newest, lastSeen: box.lastSeen)
                )
            }

            let newestRegistration =
                firestore
                .collection(Self.notificationsCollection)
                .document(uid)
                .collection(Self.itemsCollection)
                .order(by: Self.createdAtField, descending: true)
                .limit(to: 1)
                .addSnapshotListener { snapshot, error in
                    if let error {
                        // Access revoked (rules gating / restriction): hard-clear
                        // the newest instant so the dot never stays lit for an
                        // inbox the user can no longer read. A transient error
                        // WITH a cached snapshot falls through and uses it.
                        if FirebaseEventsRepository.firestoreStatusName(error) == "PERMISSION_DENIED" {
                            box.newest = nil
                            emit()
                            return
                        }
                        if snapshot == nil { return }
                    }
                    box.newest = (snapshot?.documents.first?.get(Self.createdAtField) as? Timestamp)?
                        .dateValue()
                    emit()
                }

            let markerRegistration =
                firestore
                .collection(Self.userPrivateCollection)
                .document(uid)
                .addSnapshotListener { snapshot, error in
                    // Transient failure with no cached marker: keep the
                    // last-known marker rather than momentarily reading it as
                    // missing, which could wrongly re-light the dot.
                    if error != nil, snapshot == nil { return }
                    box.lastSeen = (snapshot?.get(Self.lastSeenField) as? Timestamp)?.dateValue()
                    emit()
                }

            box.registrations = [newestRegistration, markerRegistration]
            continuation.onTermination = { _ in
                box.removeAll()
            }
        }
    }

    func markRead(notificationId: String) async throws {
        _ = try await functions.call(Self.markReadCallable, payload: ["notificationId": notificationId])
    }

    func markAllRead() async throws {
        _ = try await functions.call(Self.markAllReadCallable, payload: [:])
    }

    func markSeen() async throws {
        _ = try await functions.call(Self.markSeenCallable, payload: [:])
    }

    func currentUserId() -> String? {
        Auth.auth().currentUser?.uid
    }

    // MARK: - Mapping

    /// Tolerant document → item mapping — the iOS port of Android's
    /// `DocumentSnapshot.toNotification()`. The Firestore `Timestamp` is
    /// converted to a `Date` HERE (the Firebase-specific step) so the pure
    /// ``AppNotification/decode(id:fields:)`` decoder stays testable with a
    /// plain dictionary.
    static func notification(from document: DocumentSnapshot) -> AppNotification? {
        guard document.exists else { return nil }
        var fields = document.data() ?? [:]
        if let timestamp = document.get(createdAtField) as? Timestamp {
            fields[createdAtField] = timestamp.dateValue()
        } else {
            // A non-Timestamp value must not reach the pure decoder as a raw
            // Timestamp it cannot read; drop it so createdAt degrades to nil.
            fields[createdAtField] = nil
        }
        return AppNotification.decode(id: document.documentID, fields: fields)
    }

    // MARK: - Factory

    private static let notificationsCollection = "notifications"
    private static let itemsCollection = "items"
    private static let createdAtField = "createdAt"
    private static let userPrivateCollection = "userPrivate"
    private static let lastSeenField = "notificationsLastSeenAt"
    private static let markReadCallable = "notifications-markRead"
    private static let markAllReadCallable = "notifications-markAllRead"
    private static let markSeenCallable = "notifications-markSeen"

    private static let cachedLock = NSLock()
    nonisolated(unsafe) private static var cached: FirebaseNotificationsRepository?

    /// Returns the process-wide repository when Firebase is configured for
    /// this build, or nil when GoogleService-Info.plist is absent (CI, local
    /// validation builds — see apps/ios/README.md). Points the Firestore SDK
    /// at the emulator when `FIREBASE_FIRESTORE_EMULATOR_HOST` is set, the
    /// same seam ``FirebaseEventsRepository/createIfAvailable()`` provides.
    static func createIfAvailable() -> NotificationsRepository? {
        guard FirebaseApp.app() != nil, let functions = KccFunctionsClient.createIfAvailable() else {
            return nil
        }
        cachedLock.lock()
        defer { cachedLock.unlock() }
        if let cached { return cached }
        let firestore = Firestore.firestore()
        // The host check makes a second application of the SAME emulator
        // settings a no-op instead of mutating settings after Firestore has
        // already started (`Firestore.firestore()` is a shared singleton, and
        // other repositories' `createIfAvailable()` factories apply the same
        // settings) — the same guard `FirebaseVehiclesRepository` /
        // `FirebaseUserProfileRepository` use.
        if let emulator = FirebaseEmulatorHost.parse(
            ProcessInfo.processInfo.environment["FIREBASE_FIRESTORE_EMULATOR_HOST"]
        ), firestore.settings.host != "\(emulator.host):\(emulator.port)" {
            firestore.useEmulator(withHost: emulator.host, port: emulator.port)
        }
        let repository = FirebaseNotificationsRepository(firestore: firestore, functions: functions)
        cached = repository
        return repository
    }
}

/// `ListenerRegistration` is not Sendable, but the stream's `onTermination`
/// closure must be — all it does is remove the listener, which Firestore
/// documents as thread-safe, so the wrapper is sound (the same wrapper
/// ``FirebaseEventsRepository`` uses).
private struct ListenerBox: @unchecked Sendable {
    let registration: ListenerRegistration
}

/// Holds the two ``FirebaseNotificationsRepository/unread(uid:)`` listeners'
/// last values and registrations.
///
/// Firestore does not GUARANTEE both listeners deliver on the same queue, so
/// `newest`/`lastSeen` are lock-protected rather than assumed
/// single-threaded — each is its own critical section, which is sufficient
/// here: a caller reads BOTH via ``FirebaseNotificationsRepository/unread(uid:)``'s
/// `emit()`, and a torn read (one field one callback behind the other)
/// self-heals on the very next emission from whichever listener fired.
private final class UnreadListenerBox: @unchecked Sendable {
    private let lock = NSLock()
    private var storedNewest: Date?
    private var storedLastSeen: Date?
    private var storedRegistrations: [ListenerRegistration] = []

    var newest: Date? {
        get { lock.withLock { storedNewest } }
        set { lock.withLock { storedNewest = newValue } }
    }

    var lastSeen: Date? {
        get { lock.withLock { storedLastSeen } }
        set { lock.withLock { storedLastSeen = newValue } }
    }

    var registrations: [ListenerRegistration] {
        get { lock.withLock { storedRegistrations } }
        set { lock.withLock { storedRegistrations = newValue } }
    }

    func removeAll() {
        let registrations = lock.withLock { () -> [ListenerRegistration] in
            let current = storedRegistrations
            storedRegistrations = []
            return current
        }
        for registration in registrations {
            registration.remove()
        }
    }
}
